### PR TITLE
release-22.2: sql/importer: wait for any pending ingestion to complete before rolling back

### PR DESCRIPTION
**Pull request by dt**
Backport 3/3 commits from #97071.

/cc @cockroachdb/release

---

This change adds a new method the job registry, allowing any ingestion
activity related to a given job ID to be indicated to the job registry,
including ingestion activity that is being performed by a processor that
is being run on this node on behalf of a job being coordinated by some
other node.

The registry also exposes a method to query if a given job has indicated
to the registry on a given node that it is engaged in ingestion activity
on that node.

In conjunction, these APIs can allow a job to indicate when and where it
is running ingestion tasks and for inspection of this information by
other tasks, including potentially subsequent retries or phases of the
same job. Implementation of such usages is left to a future change.

This adds a new helper utility that can be used to check if a given job
has indicated that it is currently performing ingestion work on any node
using the job Registry's MarkAsIngesting API.

This is done by running a distSQL flow on all nodes that invokes a new
processor which simply reads this registry API, returning an error if
the given job ID is ingesting and no error if it is not.

This utility is intended for use when a job needs to positively verify
that ingestion activity for a given job -- typically itself in a previous
phase or retry -- has ceases before moving on to a new phase or retry.

Typically, a job only moves on to a new phase after the flow running its
ingestion activity returns after every ingesting processor exits. However
in some rare edge cases, it is possible for this not to be the case: a
very slow, delayed ingestion could still be flushing after the coordinator
for the job terminated and the job was adopted on a different node. In
such a case, that new execution of the job, potentially now in a reverting
phase, could race with this delayed ingestion. In the case of a revert,
such a race could lead to incomplete cleanup by the revert, as data may
be added by the delayed ingestion after the reverting phase processes a
given span.

Integration of this new utility into jobs that require it is left to a
future change.

Release note (bug fix): Fix a potential bug whereby a failed or
cancelled IMPORT could in some cases leave some of the imported rows
behind after it was cancelled, in the rare event that the writing
processes were slow enough to continue writing after the cleanup process
started. 

Epic: none.

Release justification: bug fix.

### CODE REVIEW

**Review by yuzefovich — APPROVED** [review]
A couple of nits from #101443 also apply.

> Reviewed 2 of 2 files at r1, 8 of 8 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101444#-:-NT16r1d4PktDh9zFn5r:b-yijydc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin and @dt)

___
*[`pkg/jobs/registry.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101444#-NT16S8T5eu1yNXFO2Mi:-NT16S8T5eu1yNXFO2Mj:brrwaei) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fbc12e17967e577816b3ad75261f16568e2b7ec/pkg/jobs/registry.go#L162)):*
> ```Go
> 		// access this to make per-job decisions about what to
> 		// do.
> 		draining bool
> ```

nit: this seems like not being used in this PR.

File: pkg/sql/execinfrapb/processors.proto [hunk 1]
```diff
@@ -123,9 +123,10 @@ message ProcessorCoreUnion {
   optional IndexBackfillMergerSpec indexBackfillMerger = 38;
   optional TTLSpec ttl = 39;
   optional CloudStorageTestSpec cloudStorageTest = 42;
+  optional IngestStoppedSpec ingestStopped = 44;
```
Comment (msbutler) [inline pkg/sql/execinfrapb/processors.proto:126]: any reason you skipped 43? Probably doesn't matter.

**Review by msbutler — APPROVED** [review]

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101444)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @dt, and @msbutler)

___
*[`pkg/sql/execinfrapb/processors.proto` line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101444#gh-1175505692:-NTo17gP-PiLX2Y3fYRU:buqnnfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ccae5b63e24400c4aad7d3d7164bdcdc064ba5c/pkg/sql/execinfrapb/processors.proto#L126)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

any reason you skipped 43? Probably doesn't matter.
</blockquote></details>

It does matter - there is another processor `InsertSpec` on newer versions using 43 that won't backported, so that ID is reserved (needed for compatibility in mixed-version clusters).

**Review by msbutler — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101444)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @dt, and @yuzefovich)

___
*[`pkg/sql/execinfrapb/processors.proto` line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101444#gh-1175505692:-NTo2PLM2oK0Tqfhqard:b-2f0s4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ccae5b63e24400c4aad7d3d7164bdcdc064ba5c/pkg/sql/execinfrapb/processors.proto#L126)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It does matter - there is another processor `InsertSpec` on newer versions using 43 that won't backported, so that ID is reserved (needed for compatibility in mixed-version clusters).
</blockquote></details>

huh, but there is no code in this file that reserves this id. and before david's change, the next ID was 43. I suppose new processors are rarely backported, so this isn't a vulnerability we need to worry about much.

**Review by msbutler — APPROVED** [review]

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101444)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin and @msbutler)

___
*[`pkg/sql/execinfrapb/processors.proto` line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101444#gh-1175505692:-NTsEuwN-Nr1UF35yEXh:b-713dau) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ccae5b63e24400c4aad7d3d7164bdcdc064ba5c/pkg/sql/execinfrapb/processors.proto#L126)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

huh, but there is no code in this file that reserves this id. and before david's change, the next ID was 43. I suppose new processors are rarely backported, so this isn't a vulnerability we need to worry about much.
</blockquote></details>

What @yuzefovich said -- you *never* change the field IDs between versions for a few reasons: 1) you might backport whoever is using 43 on master later b) in the mixed version case, it is very important that 22.2.x and 23.1.x be using the same id for the same type, or ignoring it if they don't know about the type. You _really_ don't want 22.2 sending 23.1 a _different_ type in field 43.

Master is always the single source of truth for what fields have what IDs, what the next ID is, etc. You _never_ author an original .proto change (including hand editing one to use different IDs) on a release branch -- you always pick your IDs on master and then backport them unchanged.

**Review by dt — COMMENTED** [review]
> nits from #101443 also apply.
Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101444#-:-NTsH6uM0z00NjOvQFO-:b-7motlk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @msbutler, and @yuzefovich)

___
*[`pkg/jobs/registry.go` line 162 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/101444#-NT16S8T5eu1yNXFO2Mi:-NTsH9d014P1BzuS7gnD:b6ts8nv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fbc12e17967e577816b3ad75261f16568e2b7ec/pkg/jobs/registry.go#L162)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this seems like not being used in this PR.
</blockquote></details>

Done.

Nice catch.
___
*[`pkg/sql/execinfrapb/processors.proto` line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/101444#gh-1175505692:-NTsHCLQ39Ruf5jTa7yL:b-d7ncr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ccae5b63e24400c4aad7d3d7164bdcdc064ba5c/pkg/sql/execinfrapb/processors.proto#L126)):*
> there is no code in this file that reserves this id. 

Exactly, _this_ is _why_ we only pick new / reserve old / mess with IDs on master, since the release branches have incomplete information.

**Pull request merged**